WIP Button component

rgarcia opened the pull request:
I took the Button that @jacobsimon and @kshen0 have in the portal2 branch of the launchpad and extended it based off of @CarlNelson's spec (https://www.dropbox.com/s/pa4b0pwshg8fp1p/Button%20Spec.png?dl=0).

I also made a really quick/hacky index.html file that runs a `Demo` component that sort of showcases the `Button`. I added a webpack config that compiles all of this, and also mirrored this branch on a `gh-pages` branch. You should be able to see the result here http://clever.github.io/components/

Some todos:
- [ ] @CarlNelson in person review
- [x] @jacobsimon review on gh-pages/webpack setup, since you're thinking about this
- [ ] @rgarcia document `props` in readme
- [x] @rgarcia implement transitions from spec
- [x] Address active state issue @jacobsimon brought up
- [x] Webpack-ify Clever font

cc @Clever/front-end

eighnjel commented:
They look like tabs o.0 ... but it matches the design spec so ¯\_(ツ)_/¯

jacobsimon commented:
Nice! I think the font-size is broken. Have you tried importing this in a project? I'm not sure if having another webpack config will cause problems or not.

osdiab commented:
@jacobsimon screenshots:

![screen shot 2016-03-29 at 11 29 24 am](https://cloud.githubusercontent.com/assets/1890926/14119135/927d9f4a-f5a1-11e5-8a2e-6fe4e6cc1a0b.png)
![screen shot 2016-03-29 at 11 29 29 am](https://cloud.githubusercontent.com/assets/1890926/14119134/927d3046-f5a1-11e5-8fa4-dd89c7ba40fe.png)

osdiab commented:
@jacobsimon i think that should be fine given that `index.js` is listed as `main`, but it's worth checking.

jacobsimon commented on src/Button/Button.less:
```diff
@@ -0,0 +1,75 @@
+  &:hover,
+  &:focus,
+  &:active {
+    //outline: 0;
```
This will get rid of the blue border when you click/focus, which is fine in terms of accessibility since we're also changing the color. We should probably add a different active state since there isn't any feedback when you actually click, but I'll leave that to carl

cozmo replied:
Note - I did this as part of 8a16bcd

jacobsimon commented on src/Button/Button.jsx:
```diff
@@ -0,0 +1,23 @@
+      classes += `Button--regular `;
+    }
+    return (
+      <button className={classes} onClick={this.props.onClick} disabled={this.props.disabled || false}>
```
Could we change this to render `<a>` if `href` is provided?

osdiab replied:
make sure it's accessible :3

jacobsimon commented on .babelrc:
```diff
@@ -0,0 +1,6 @@
+{
+  "presets": [
+    "es2015",
```
Using es6+ here will prevent us from using this in any repo that hasn't been converted yet, unless we're going to compile the components down to es5 and then import them. Not necessarily bad but something to keep in mind.

rgarcia replied:
AFAIK ES6 == ES2015, so this is just E6, but yeah, this might make it harder to import in other repos. I'd like to first try importing `clever/components` in another repo before settling for less than ES6.

kshen0 commented on package.json:
```diff
@@ -15,11 +15,21 @@
+    "webpack-dev-server": "^1.14.1"
+  },
+  "scripts": {
+    "dev-server": "node_modules/.bin/webpack-dev-server --watch --host 192.168.50.4"
```
What host is this? Is this intentionally hardcoded?

rgarcia replied:
this is the VM IP...this is the only way I could get the webpack dev server to load on my host machine's browser :-/

osdiab replied:
you should be able to forward your VM ports in such a way that this isn't necessary—but that's not for this PR, so I say we should remove the `--host` argument.

rgarcia replied:
yeah, I tried that (in many different variations) and nothing worked

osdiab commented:
doesn't the spec specify ease-in transitions?

osdiab commented on index.html:
```diff
@@ -0,0 +1,7 @@
+<!DOCTYPE html>
```
If this is just for demo purposes, can we not put it in the project root? Seems counterintuitive to be seeing stuff like this when you first see the project. Same goes for webpack config if it's just for demo purposes.

jacobsimon replied:
ideally this would live on the gh-pages branch

rgarcia replied:
I'll give it a shot... so far `webpack-dev-server` has been really frustrating/difficult to get right, so I've tried to keep it simple.

rgarcia commented:
@osdiab good catch, will add as a todo

osdiab commented on src/Button/Button.less:
```diff
@@ -0,0 +1,75 @@
+
+.Button--small {
+  padding: 10px 15px 5px 15px;
+  font-size: 10px semibold;
```
Pretty sure these font-size declarations don't work.

rgarcia replied:
any suggestions for how to fix them? From what I can tell the user agent stylesheet `font-size` is taking precedence, and I have no idea why.

eighnjel replied:
i think they need to be separate declarations:
```
font-size: 10px;
font-weight: semibold;
```

rgarcia replied:
:boom: that worked, thanks @eighnjel!

eighnjel replied:
If you want to do shorthand, you should use `font` instead of `font-size` and need to specify at least font-size and font-family:
```
font: 10px "some-font", semibold;
```

osdiab commented on src/Button/Button.jsx:
```diff
@@ -0,0 +1,23 @@
+
+require('./Button.less');
+
+class Button extends React.Component {
```
I know this is simple but can we have doc comments to explain what these actually are, what props they can take, what state they have, etc?

Additionally, maybe this would be a good time to actually use [React `propTypes`](https://facebook.github.io/react/docs/reusable-components.html).

rgarcia replied:
@osdiab I already have a TODO to add docs to the README--are you suggesting additional documentation as comments in the code? I'd be worried about keeping these in sync.

Agreed that we should add `propTypes`

rgarcia commented:
made some updates here: 272b6ad

- Put most gh-pages related stuff in docs/. There's still a webpack config and package.json in the root of the repo since moving these to sub-folders was a world of pain
- Added a deploy script for publishing the docs/ subfolder to the gh-pages branch
- Added support for buttons-as-links
- `npm run-script dev-server` runs the gh pages site w/ live reload. I also removed hardcoded VM IP (someone should try `npm run-script dev-server` and see if they can load localhost:8080, because I can't)

Still have TODOs listed in the PR description

rgarcia commented:
cc @cozmo  added a link to the spec in the PR description, and also a TODO to maybe figure out fonts

jacobsimon commented on src/Button/Button.jsx:
```diff
@@ -0,0 +1,43 @@
+    let size = ["large", "regular", "small"].indexOf(this.props.size) > -1 ? this.props.size : "regular";
+    classes += `Button--${size} `;
+    if (this.props.href == null && !this.props.disabled) {
+      // use <button>s for all disabled buttons and things with no href prop
```
This is currently using links when disabled instead of buttons

jacobsimon commented on src/Button/Button.jsx:
```diff
@@ -0,0 +1,43 @@
+require('./Button.less');
+
+const VALID_TYPES = ["primary", "secondary", "destructive"];
+const VALID_SIZES = ["large", "regular", "small", "regular"];
```
regular is listed twice!

jacobsimon commented on src/Button/Button.jsx:
```diff
@@ -0,0 +1,43 @@
+export class Button extends React.Component {
+  constructor(props) {
+    super();
+    if (props.type && VALID_TYPES.indexOf(props.type) === -1) {
```
We could get most of this validation for free/cheap if we declare propTypes like @osdiab mentioned. For example:
```javascript
Button.propTypes = {
  type: React.PropTypes.oneOf(['primary', 'secondary', 'destructive']).isRequired,
  size: React.PropTypes.oneOf(['large', 'regular', 'small']),
  value: React.PropTypes.string.isRequired,
  disabled: React.PropTypes.bool
}
```

rgarcia replied:
cc @cozmo

jacobsimon commented on src/Button/Button.jsx:
```diff
@@ -0,0 +1,43 @@
+    if (this.props.href == null && !this.props.disabled) {
+      // use <button>s for all disabled buttons and things with no href prop
+      return (
+        <button className={classes} onClick={this.props.onClick} disabled={this.props.disabled || false}>
```
We don't need `|| false` if we assume `props.disabled` is bool

jacobsimon commented on src/Button/Button.less:
```diff
@@ -0,0 +1,89 @@
+  &.Button--small {
+    padding: 10px 15px 5px 15px;
+    font-size: 10px;
+    font-weight: semibold;
```
don't think this is a valid [font-weight](https://developer.mozilla.org/en-US/docs/Web/CSS/font-weight), probably want `600`

jacobsimon commented on src/Button/Button.jsx:
```diff
@@ -0,0 +1,43 @@
+        </button>
+      );
+    } else {
+      const target = this.props.target != null ? this.props.target : "_self";
```
we _could_ also use `getDefaultProps` for target, disabled, and any other optional params

jacobsimon commented on src/Button/Button.jsx:
```diff
@@ -0,0 +1,43 @@
+      );
+    } else {
+      const target = this.props.target != null ? this.props.target : "_self";
+      const onClick = this.props.disabled ? function(e){ e.preventDefault(); } : null;
```
this works but you could also just remove the href if it's disabled, or use `pointer-events: none` in the styles

rgarcia replied:
cc @cozmo

jacobsimon commented on package.json:
```diff
@@ -15,11 +15,21 @@
+    "webpack-dev-server": "^1.14.1"
+  },
+  "scripts": {
+    "dev-server": "node_modules/.bin/webpack-dev-server --content-base docs/ --inline --watch"
```
Do we need webpack-dev-server?

rgarcia replied:
`webpack-dev-server --inline` adds a bit of js into the bundle that detects changes and reloads the page for you. So far I've found it a lot better than `webpack --watch` + opening `index.html` + reloading manually.

cozmo commented:
@jacobsimon I pushed some changes (555fab7) that took a lot of your comments into account. I think it's looking pretty good now. Thanks!

@rgarcia and I are going to try to get the fonts working next.
